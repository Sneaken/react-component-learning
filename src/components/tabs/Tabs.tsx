import { CloseOutlined, EllipsisOutlined, PlusOutlined } from '@ant-design/icons';
import classnames from 'classnames';
import {
  type CSSProperties,
  forwardRef,
  type HTMLAttributes,
  type KeyboardEvent,
  type MouseEvent,
  type ReactNode,
  type Ref,
  useEffect,
  useMemo,
  useState,
} from 'react';
import useMergedState from '../../utils/hooks/useMergedState';
import useAnimateConfig from './hooks/useAnimateConfig';
import type { AnimatedConfig, EditableConfig, RenderTabBar, Tab, TabBarExtraContent } from './interface';
import { OnTabScroll } from './interface';
import './style/index.css';
import TabNavListWrapper from './tab-nav-list/wrapper';
import TabPanelList from './tab-panel-list';
import TabContext from './TabContext';

export type TabsType = 'line' | 'card' | 'editable-card';
export type SizeType = 'small' | 'middle' | 'large';

export interface TabsProps extends Omit<HTMLAttributes<HTMLDivElement>, 'onChange' | 'children'> {
  activeKey?: string;
  addIcon?: ReactNode;
  animated?: boolean | AnimatedConfig;
  centered?: boolean;
  defaultActiveKey?: string;
  destroyInactiveTabPane?: boolean;
  direction?: 'ltr' | 'rtl';
  getPopupContainer?: (node: HTMLElement) => HTMLElement;
  hideAdd?: boolean;
  items?: Tab[];
  onChange?: (activeKey: string) => void;
  onEdit?: (e: MouseEvent | KeyboardEvent | string, action: 'add' | 'remove') => void;
  onTabClick?: (key: string, event: MouseEvent) => void;
  popupClassName?: string;
  renderTabBar?: RenderTabBar;
  size?: SizeType;
  tabBarGutter?: number;
  tabBarStyle?: CSSProperties;
  tabPosition?: 'top' | 'bottom' | 'left' | 'right';
  type?: TabsType;

  moreIcon?: ReactNode;
  onTabScroll?: OnTabScroll;
  tabBarExtraContent?: TabBarExtraContent;
}

function Tabs(props: TabsProps, ref: Ref<HTMLDivElement>) {
  const {
    activeKey,
    addIcon,
    animated = { inkBar: true, tabPane: false },
    centered = false,
    className,
    defaultActiveKey,
    destroyInactiveTabPane,
    direction = 'ltr',
    getPopupContainer = () => document.body,
    hideAdd,
    items,
    onChange,
    onEdit,
    onTabClick,
    popupClassName,
    renderTabBar,
    size,
    tabBarGutter,
    tabBarStyle,
    tabPosition = 'top',
    type = 'line',
    moreIcon = <EllipsisOutlined />,
    onTabScroll,
    tabBarExtraContent,
    ...restProps
  } = props;
  const tabs = useMemo(
    () => (items || []).filter((item) => item && typeof item === 'object' && Object.hasOwn(item, 'key')),
    [items]
  );
  const rtl = direction === 'rtl';

  const mergedAnimated = useAnimateConfig(animated);

  let editable: EditableConfig | undefined;
  if (type === 'editable-card') {
    editable = {
      onEdit: (editType, { key, event }) => {
        onEdit?.(editType === 'add' ? event : key!, editType);
      },
      removeIcon: <CloseOutlined />,
      addIcon: addIcon || <PlusOutlined />,
      showAdd: hideAdd !== true,
    };
  }

  // ====================== Active Key ======================
  const [mergedActiveKey, setMergedActiveKey] = useMergedState<string>(() => tabs[0]?.key, {
    value: activeKey,
    defaultValue: defaultActiveKey,
  });
  const [activeIndex, setActiveIndex] = useState(() => tabs.findIndex((tab) => tab.key === mergedActiveKey));

  // Reset active key if not exist anymore
  useEffect(() => {
    let newActiveIndex = tabs.findIndex((tab) => tab.key === mergedActiveKey);
    if (newActiveIndex === -1) {
      newActiveIndex = Math.max(0, Math.min(activeIndex, tabs.length - 1));
      setMergedActiveKey(tabs[newActiveIndex]?.key);
    }
    setActiveIndex(newActiveIndex);
  }, [tabs.map((tab) => tab.key).join('_'), mergedActiveKey, activeIndex]);

  // ======================== Events ========================
  function onInternalTabClick(key: string, e: MouseEvent | KeyboardEvent) {
    onTabClick?.(key, e as MouseEvent);
    const isActiveChanged = key !== mergedActiveKey;
    setMergedActiveKey(key);
    isActiveChanged && onChange?.(key);
  }

  // ======================== Render ========================
  const sharedProps = {
    activeKey: mergedActiveKey,
    animated: mergedAnimated,
    rtl,
    tabPosition,
  };
  const tabNavBarProps = {
    ...sharedProps,
    editable,
    extra: tabBarExtraContent,
    getPopupContainer,
    moreIcon,
    moreTransitionName: 'slide-up',
    onTabClick: onInternalTabClick,
    onTabScroll,
    panes: null,
    popupClassName,
    style: tabBarStyle,
    tabBarGutter,
  };
  return (
    <TabContext.Provider value={{ tabs }}>
      <div
        className={classnames(
          'tabs',
          `tabs-${tabPosition}`,
          {
            [`tabs-${size}`]: size,
            [`tabs-card`]: ['card', 'editable-card'].includes(type),
            [`tabs-centered`]: centered,
            [`tabs-editable-card`]: type === 'editable-card',
            [`tabs-editable`]: !!editable,
            [`tabs-rtl`]: rtl,
          },
          className
        )}
        ref={ref}
        {...restProps}
      >
        <TabNavListWrapper {...tabNavBarProps} renderTabBar={renderTabBar} />
        <TabPanelList destroyInactiveTabPane={destroyInactiveTabPane} {...sharedProps} animated={mergedAnimated} />
      </div>
    </TabContext.Provider>
  );
}

export default forwardRef(Tabs);
