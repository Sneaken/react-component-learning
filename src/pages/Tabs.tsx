import React, { forwardRef, Ref, useState } from 'react';
import { Tabs as AntdTab } from 'antd';
import { ResizeObserver, Tabs } from '../components';

const Label = forwardRef((props, ref: Ref<any>) => {
  return <p ref={ref}>222</p>;
});

const items = [
  {
    label: <Label />,
    key: '1',
    className: '222',
    children: `Content of Tab Pane 1`,
  },
  {
    label: `Tab 2`,
    key: '2',
    children: `Content of Tab Pane 2`,
  },
  {
    label: `Tab 3`,
    key: '3',
    children: `Content of Tab Pane 3`,
  },
];
const PageOfTabs = () => {
  const [activeKey, setActiveKey] = useState<string>();
  const onChange = (key: string) => {
    setActiveKey(key);
  };

  return (
    <>
      <div style={{ padding: '0 20px' }}>
        <div style={{ float: 'left', width: '40vw' }}>
          <Tabs activeKey={activeKey} onChange={onChange} items={items} />
        </div>
        <div style={{ float: 'left', width: '40vw' }}>
          <AntdTab activeKey={activeKey} onChange={onChange} items={items} />
        </div>
      </div>
      <div>
        <ResizeObserver>
          <Label />
        </ResizeObserver>
        <ResizeObserver>
          {(ref) => {
            return <Label ref={ref}></Label>;
          }}
        </ResizeObserver>
      </div>
    </>
  );
};

export default PageOfTabs;
