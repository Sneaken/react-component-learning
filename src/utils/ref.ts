import { type Ref, useRef } from 'react';
import { isMemo } from 'react-is';

export function fillRef<T>(ref: Ref<T>, node: T) {
  if (typeof ref === 'function') {
    ref(node);
  } else if (typeof ref === 'object' && ref && Object.hasOwn(ref, 'current')) {
    // MutableRefObject
    (ref as any).current = node;
  }
}

/**
 * Merge refs into one ref function to support ref passing.
 */
export function composeRef<T>(...refs: Ref<T>[]): Ref<T> {
  const refList = refs.filter((ref) => ref);
  if (refList.length <= 1) return refList[0];

  return (node: T) => {
    refs.forEach((ref) => {
      fillRef(ref, node);
    });
  };
}

interface Cache<Value, Condition> {
  condition?: Condition;
  value?: Value;
}

export function useComposeRef<T>(...refs: Ref<T>[]): Ref<T> | undefined {
  const cacheRef = useRef<Cache<Ref<T>, Ref<T>[]>>({});
  const shouldUpdate = (prev: Ref<T>[], next: Ref<T>[]) =>
    prev.length === next.length && prev.every((ref, i) => ref === next[i]);

  if (!Object.hasOwn(cacheRef.current, 'value') || shouldUpdate(cacheRef.current.condition!, refs)) {
    cacheRef.current.value = composeRef(...refs);
    cacheRef.current.condition = refs;
  }

  return cacheRef.current.value;
}

export function supportRef(nodeOrComponent: any): boolean {
  const type = isMemo(nodeOrComponent) ? nodeOrComponent.type.type : nodeOrComponent.type;
  // Function component node 不是 forwardRef 包裹的时候是不能挂载 ref 的
  if (typeof type === 'function' && !type.prototype?.render) return false;
  // Class component 没有 render 函数也是一样
  if (typeof nodeOrComponent === 'function' && !nodeOrComponent.prototype?.render) return false;
  return true;
}
