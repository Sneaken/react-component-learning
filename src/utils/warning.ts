import noop from '@/utils/noop';

type Warning = (valid: boolean, component: string, message?: string) => void;

export let warning: Warning = noop;

if (process.env.NODE_ENV !== 'production') {
  warning = (valid, component, message) => {
    if (!valid && console !== undefined) {
      console.error(`Warning: [${component}] ${message}`);
    }
  };
}
