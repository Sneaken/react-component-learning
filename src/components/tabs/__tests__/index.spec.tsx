import KeyCode from '@/utils/KeyCode';
import { spyElementPrototypes } from '@tests/domHook';
import { getOffsetSizeFunc, HackInfo } from './utils';
import mountTest from '@tests/shared/mountTest';
import { fireEvent, render, waitFakeTimer } from '@tests/utils';
import { Mock } from 'vitest';
import Tabs, { Tab, TabsProps } from '..';

describe('Tabs', () => {
  let domSpy: ReturnType<typeof spyElementPrototypes>;

  const hackOffsetInfo: HackInfo = {};

  beforeEach(() => {
    Object.keys(hackOffsetInfo).forEach((key) => {
      delete hackOffsetInfo[key as keyof HackInfo];
    });
  });

  beforeAll(() => {
    domSpy = spyElementPrototypes(HTMLElement, {
      scrollIntoView: () => {},
      offsetWidth: {
        get: getOffsetSizeFunc(hackOffsetInfo),
      },
    });
  });

  afterAll(() => {
    domSpy.mockRestore();
  });

  mountTest(() => (
    <Tabs
      items={[
        {
          key: 'xx',
          label: 'xx',
        },
      ]}
    />
  ));

  describe('editable-card', () => {
    let handleEdit: Mock;
    let wrapper: ReturnType<typeof render>['container'];
    const key = 'foo';

    beforeEach(() => {
      handleEdit = vi.fn();
      const { container } = render(
        <Tabs
          type="editable-card"
          onEdit={handleEdit}
          items={[
            {
              label: 'foo',
              key,
            },
          ]}
        />
      );
      wrapper = container;
    });

    it('add card', () => {
      fireEvent.click(wrapper.querySelector('.tabs-nav-add')!);
      expect(handleEdit.mock.calls[0][1]).toBe('add');
    });

    it('remove card', () => {
      fireEvent.click(wrapper.querySelector('.anticon-close')!);
      expect(handleEdit).toHaveBeenCalledWith(key, 'remove');
    });

    it('validateElement', () => {
      expect(wrapper.querySelectorAll('.tabs-tab')).toHaveLength(1);
    });
  });

  describe('tabPosition', () => {
    it('position in left', () => {
      const { container } = render(
        <Tabs
          tabPosition="left"
          items={[
            {
              label: 'foo',
              key: '1',
              children: 'foo',
            },
          ]}
        />
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('position in right', () => {
      const { container } = render(
        <Tabs
          tabPosition="right"
          items={[
            {
              label: 'foo',
              key: '1',
              children: 'foo',
            },
          ]}
        />
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('position in top', () => {
      const { container } = render(
        <Tabs
          tabPosition="top"
          items={[
            {
              label: 'foo',
              key: '1',
              children: 'foo',
            },
          ]}
        />
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('position in bottom', () => {
      const { container } = render(
        <Tabs
          tabPosition="bottom"
          items={[
            {
              label: 'foo',
              key: '1',
              children: 'foo',
            },
          ]}
        />
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('renderTabBar', () => {
    it('custom-tab-bar', () => {
      const { container } = render(
        <Tabs
          items={[
            {
              label: 'foo',
              key: '1',
            },
          ]}
          renderTabBar={() => <div>custom-tab-bar</div>}
        />
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  it('tabBarGutter should work', () => {
    const items: Tab[] = Array.from({ length: 3 }).map((_, i) => {
      return {
        label: i,
        key: i,
        children: i,
      };
    }) as any;
    const { container: wrapper } = render(<Tabs items={items} tabBarGutter={4} />);
    expect(wrapper.firstChild).toMatchSnapshot();
    const { container: wrapper2 } = render(<Tabs items={items} tabBarGutter={4} tabPosition="left" />);
    expect(wrapper2.firstChild).toMatchSnapshot();
  });

  function getTabs(props: TabsProps = {}) {
    const mergedProps = {
      items: [
        {
          label: 'light',
          key: 'light',
          children: 'Light',
        },
        {
          label: 'bamboo',
          key: 'bamboo',
          children: 'Bamboo',
        },
        {
          label: 'cute',
          key: 'cute',
          children: 'Cute',
        },
      ],
      ...props,
    };

    return <Tabs {...mergedProps} />;
  }

  it('disabled not change', () => {
    const onChange = vi.fn();

    const { container } = render(
      getTabs({
        defaultActiveKey: 'light',
        items: [
          {
            label: 'light',
            key: 'light',
            children: 'Light',
          },
          {
            label: 'disabled',
            key: 'disabled',
            children: 'Disabled',
            disabled: true,
          },
        ],
        onChange,
      })
    );

    fireEvent.click(container.querySelector('.tabs-tab-disabled')!);
    expect(onChange).not.toHaveBeenCalled();
  });

  it('Skip invalidate children', () => {
    const { container } = render(
      getTabs({
        items: [
          {
            label: 'light',
            key: 'light',
            children: 'Light',
          },
          'not me' as any,
        ],
      })
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('strictMode should show correct ink bar', () => {
    vi.useFakeTimers();

    const { container } = render(
      <Tabs
        items={new Array(2).fill(0).map((_, index) => ({
          label: `Tab ${index}`,
          key: `${index}`,
          children: `Content of Tab Pane ${index}`,
        }))}
        activeKey="0"
      />
    );

    waitFakeTimer(0, 100);

    expect(container.querySelector<HTMLElement>('.tabs-ink-bar')?.style?.width).toBe('20px');

    vi.useRealTimers();
  });

  describe('onChange and onTabClick should work', () => {
    const list: { name: string; trigger: (container: HTMLElement) => void }[] = [
      {
        name: 'outer div',
        trigger: (container) => fireEvent.click(container.querySelectorAll('.tabs-tab')[2]),
      },
      {
        name: 'inner button',
        trigger: (container) => fireEvent.click(container.querySelectorAll('.tabs-tab .tabs-tab-btn')[2]),
      },
      {
        name: 'inner button key down',
        trigger: (container) =>
          fireEvent.keyDown(container.querySelectorAll('.tabs-tab .tabs-tab-btn')[2], {
            which: KeyCode.SPACE,
            keyCode: KeyCode.SPACE,
            charCode: KeyCode.SPACE,
          }),
      },
    ];

    list.forEach(({ name, trigger }) => {
      it(name, () => {
        const onChange = vi.fn();
        const onTabClick = vi.fn();
        const { container } = render(getTabs({ onChange, onTabClick }));

        trigger(container);
        expect(onTabClick).toHaveBeenCalledWith('cute', expect.anything());
        expect(onChange).toHaveBeenCalledWith('cute');
      });
    });

    it('should not trigger onChange when click current tab', () => {
      const onChange = vi.fn();
      const onTabClick = vi.fn();
      const { container } = render(getTabs({ onChange, onTabClick }));

      fireEvent.click(container.querySelector('.tabs-tab')!);
      expect(onTabClick).toHaveBeenCalledWith('light', expect.anything());
      expect(onChange).not.toHaveBeenCalled();
    });
  });
});
