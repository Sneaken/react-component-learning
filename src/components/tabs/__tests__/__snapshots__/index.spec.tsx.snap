// Vitest Snapshot v1

exports[`Tabs > Skip invalidate children 1`] = `
<div
  class="tabs tabs-top"
>
  <div
    class="tabs-nav"
    role="tablist"
  >
    <div
      class="tabs-nav-wrap"
    >
      <div
        class="tabs-nav-list"
        style="transform: translate(0px, 0px);"
      >
        <div
          class="tabs-tab tabs-tab-active"
          data-node-key="light"
        >
          <div
            aria-selected="true"
            class="tabs-tab-btn"
            role="tab"
            tabindex="0"
          >
            light
          </div>
        </div>
        <div
          class="tabs-ink-bar tabs-ink-bar-animated"
        />
      </div>
    </div>
    <div
      class="tabs-nav-operations tabs-nav-operations-hidden"
    >
      <button
        aria-controls="-more-popup"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-hidden="true"
        class="tabs-nav-more"
        id="-more"
        style="visibility: hidden; order: 1;"
        tabindex="-1"
        type="button"
      >
        <span
          aria-label="ellipsis"
          class="anticon anticon-ellipsis"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="ellipsis"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
  <div
    class="tabs-content-holder"
  >
    <div
      class="tabs-content tabs-content-top"
    >
      <div
        class="tabs-tabpane tabs-active"
        role="tabpanel"
        tabindex="0"
      >
        Light
      </div>
    </div>
  </div>
</div>
`;

exports[`Tabs > renderTabBar > custom-tab-bar 1`] = `
<div
  class="tabs tabs-top"
>
  <div>
    custom-tab-bar
  </div>
  <div
    class="tabs-content-holder"
  >
    <div
      class="tabs-content tabs-content-top"
    >
      <div
        class="tabs-tabpane tabs-active"
        role="tabpanel"
        tabindex="0"
      />
    </div>
  </div>
</div>
`;

exports[`Tabs > tabBarGutter should work 1`] = `
<div
  class="tabs tabs-top"
>
  <div
    class="tabs-nav"
    role="tablist"
  >
    <div
      class="tabs-nav-wrap"
    >
      <div
        class="tabs-nav-list"
        style="transform: translate(0px, 0px);"
      >
        <div
          class="tabs-tab tabs-tab-active"
          data-node-key="0"
        >
          <div
            aria-selected="true"
            class="tabs-tab-btn"
            role="tab"
            tabindex="0"
          >
            0
          </div>
        </div>
        <div
          class="tabs-tab"
          data-node-key="1"
          style="margin-left: 4px;"
        >
          <div
            aria-selected="false"
            class="tabs-tab-btn"
            role="tab"
            tabindex="0"
          >
            1
          </div>
        </div>
        <div
          class="tabs-tab"
          data-node-key="2"
          style="margin-left: 4px;"
        >
          <div
            aria-selected="false"
            class="tabs-tab-btn"
            role="tab"
            tabindex="0"
          >
            2
          </div>
        </div>
        <div
          class="tabs-ink-bar tabs-ink-bar-animated"
        />
      </div>
    </div>
    <div
      class="tabs-nav-operations tabs-nav-operations-hidden"
    >
      <button
        aria-controls="-more-popup"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-hidden="true"
        class="tabs-nav-more"
        id="-more"
        style="margin-left: 4px; visibility: hidden; order: 1;"
        tabindex="-1"
        type="button"
      >
        <span
          aria-label="ellipsis"
          class="anticon anticon-ellipsis"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="ellipsis"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
  <div
    class="tabs-content-holder"
  >
    <div
      class="tabs-content tabs-content-top"
    >
      <div
        class="tabs-tabpane tabs-active"
        role="tabpanel"
        tabindex="0"
      >
        0
      </div>
    </div>
  </div>
</div>
`;

exports[`Tabs > tabBarGutter should work 2`] = `
<div
  class="tabs tabs-left"
>
  <div
    class="tabs-nav"
    role="tablist"
  >
    <div
      class="tabs-nav-wrap"
    >
      <div
        class="tabs-nav-list"
        style="transform: translate(0px, 0px);"
      >
        <div
          class="tabs-tab tabs-tab-active"
          data-node-key="0"
        >
          <div
            aria-selected="true"
            class="tabs-tab-btn"
            role="tab"
            tabindex="0"
          >
            0
          </div>
        </div>
        <div
          class="tabs-tab"
          data-node-key="1"
          style="margin-top: 4px;"
        >
          <div
            aria-selected="false"
            class="tabs-tab-btn"
            role="tab"
            tabindex="0"
          >
            1
          </div>
        </div>
        <div
          class="tabs-tab"
          data-node-key="2"
          style="margin-top: 4px;"
        >
          <div
            aria-selected="false"
            class="tabs-tab-btn"
            role="tab"
            tabindex="0"
          >
            2
          </div>
        </div>
        <div
          class="tabs-ink-bar tabs-ink-bar-animated"
        />
      </div>
    </div>
    <div
      class="tabs-nav-operations tabs-nav-operations-hidden"
    >
      <button
        aria-controls="-more-popup"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-hidden="true"
        class="tabs-nav-more"
        id="-more"
        style="margin-left: 4px; visibility: hidden; order: 1;"
        tabindex="-1"
        type="button"
      >
        <span
          aria-label="ellipsis"
          class="anticon anticon-ellipsis"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="ellipsis"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
  <div
    class="tabs-content-holder"
  >
    <div
      class="tabs-content tabs-content-left"
    >
      <div
        class="tabs-tabpane tabs-active"
        role="tabpanel"
        tabindex="0"
      >
        0
      </div>
    </div>
  </div>
</div>
`;

exports[`Tabs > tabPosition > position in bottom 1`] = `
<div
  class="tabs tabs-bottom"
>
  <div
    class="tabs-nav"
    role="tablist"
  >
    <div
      class="tabs-nav-wrap"
    >
      <div
        class="tabs-nav-list"
        style="transform: translate(0px, 0px);"
      >
        <div
          class="tabs-tab tabs-tab-active"
          data-node-key="1"
        >
          <div
            aria-selected="true"
            class="tabs-tab-btn"
            role="tab"
            tabindex="0"
          >
            foo
          </div>
        </div>
        <div
          class="tabs-ink-bar tabs-ink-bar-animated"
        />
      </div>
    </div>
    <div
      class="tabs-nav-operations tabs-nav-operations-hidden"
    >
      <button
        aria-controls="-more-popup"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-hidden="true"
        class="tabs-nav-more"
        id="-more"
        style="visibility: hidden; order: 1;"
        tabindex="-1"
        type="button"
      >
        <span
          aria-label="ellipsis"
          class="anticon anticon-ellipsis"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="ellipsis"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
  <div
    class="tabs-content-holder"
  >
    <div
      class="tabs-content tabs-content-bottom"
    >
      <div
        class="tabs-tabpane tabs-active"
        role="tabpanel"
        tabindex="0"
      >
        foo
      </div>
    </div>
  </div>
</div>
`;

exports[`Tabs > tabPosition > position in left 1`] = `
<div
  class="tabs tabs-left"
>
  <div
    class="tabs-nav"
    role="tablist"
  >
    <div
      class="tabs-nav-wrap"
    >
      <div
        class="tabs-nav-list"
        style="transform: translate(0px, 0px);"
      >
        <div
          class="tabs-tab tabs-tab-active"
          data-node-key="1"
        >
          <div
            aria-selected="true"
            class="tabs-tab-btn"
            role="tab"
            tabindex="0"
          >
            foo
          </div>
        </div>
        <div
          class="tabs-ink-bar tabs-ink-bar-animated"
        />
      </div>
    </div>
    <div
      class="tabs-nav-operations tabs-nav-operations-hidden"
    >
      <button
        aria-controls="-more-popup"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-hidden="true"
        class="tabs-nav-more"
        id="-more"
        style="visibility: hidden; order: 1;"
        tabindex="-1"
        type="button"
      >
        <span
          aria-label="ellipsis"
          class="anticon anticon-ellipsis"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="ellipsis"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
  <div
    class="tabs-content-holder"
  >
    <div
      class="tabs-content tabs-content-left"
    >
      <div
        class="tabs-tabpane tabs-active"
        role="tabpanel"
        tabindex="0"
      >
        foo
      </div>
    </div>
  </div>
</div>
`;

exports[`Tabs > tabPosition > position in right 1`] = `
<div
  class="tabs tabs-right"
>
  <div
    class="tabs-nav"
    role="tablist"
  >
    <div
      class="tabs-nav-wrap"
    >
      <div
        class="tabs-nav-list"
        style="transform: translate(0px, 0px);"
      >
        <div
          class="tabs-tab tabs-tab-active"
          data-node-key="1"
        >
          <div
            aria-selected="true"
            class="tabs-tab-btn"
            role="tab"
            tabindex="0"
          >
            foo
          </div>
        </div>
        <div
          class="tabs-ink-bar tabs-ink-bar-animated"
        />
      </div>
    </div>
    <div
      class="tabs-nav-operations tabs-nav-operations-hidden"
    >
      <button
        aria-controls="-more-popup"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-hidden="true"
        class="tabs-nav-more"
        id="-more"
        style="visibility: hidden; order: 1;"
        tabindex="-1"
        type="button"
      >
        <span
          aria-label="ellipsis"
          class="anticon anticon-ellipsis"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="ellipsis"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
  <div
    class="tabs-content-holder"
  >
    <div
      class="tabs-content tabs-content-right"
    >
      <div
        class="tabs-tabpane tabs-active"
        role="tabpanel"
        tabindex="0"
      >
        foo
      </div>
    </div>
  </div>
</div>
`;

exports[`Tabs > tabPosition > position in top 1`] = `
<div
  class="tabs tabs-top"
>
  <div
    class="tabs-nav"
    role="tablist"
  >
    <div
      class="tabs-nav-wrap"
    >
      <div
        class="tabs-nav-list"
        style="transform: translate(0px, 0px);"
      >
        <div
          class="tabs-tab tabs-tab-active"
          data-node-key="1"
        >
          <div
            aria-selected="true"
            class="tabs-tab-btn"
            role="tab"
            tabindex="0"
          >
            foo
          </div>
        </div>
        <div
          class="tabs-ink-bar tabs-ink-bar-animated"
        />
      </div>
    </div>
    <div
      class="tabs-nav-operations tabs-nav-operations-hidden"
    >
      <button
        aria-controls="-more-popup"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-hidden="true"
        class="tabs-nav-more"
        id="-more"
        style="visibility: hidden; order: 1;"
        tabindex="-1"
        type="button"
      >
        <span
          aria-label="ellipsis"
          class="anticon anticon-ellipsis"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="ellipsis"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
  <div
    class="tabs-content-holder"
  >
    <div
      class="tabs-content tabs-content-top"
    >
      <div
        class="tabs-tabpane tabs-active"
        role="tabpanel"
        tabindex="0"
      >
        foo
      </div>
    </div>
  </div>
</div>
`;
